src/DebugSql.php

```diff
@@ -0,0 +1,188 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         DebugKit 3.11.4
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace DebugKit;
+
+use Cake\Core\Configure;
+use Cake\Error\Debugger;
+use Cake\ORM\Query;
+use SqlFormatter;
+
+/**
+ * Contains methods for dumping well formatted SQL queries.
+ */
+class DebugSql
+{
+    /**
+     * Template used for HTML output.
+     *
+     * @var string
+     */
+    private static $templateHtml = <<<HTML
+<div class="cake-debug-output">
+%s
+<pre class="cake-debug">
+%s
+</pre>
+</div>
+HTML;
+
+    /**
+     * Template used for CLI and text output.
+     *
+     * @var string
+     */
+    private static $templateText = <<<TEXT
+%s
+########## DEBUG ##########
+%s
+###########################
+
+TEXT;
+
+    /**
+     * Prints out the SQL statements generated by a Query object.
+     *
+     * This function returns the same variable that was passed.
+     * Only runs if debug mode is enabled.
+     *
+     * @param Query $query Query to show SQL statements for.
+     * @param bool $showValues Renders the SQL statement with bound variables.
+     * @param bool|null $showHtml If set to true, the method prints the debug
+     *    data in a browser-friendly way.
+     * @param int $stackDepth Provides a hint as to which file in the call stack to reference.
+     * @return Query
+     */
+    public static function sql(Query $query, $showValues = true, $showHtml = null, $stackDepth = 0)
+    {
+        if (!Configure::read('debug')) {
+            return $query;
+        }
+
+        $sql = (string)$query;
+        if ($showValues) {
+            $sql = method_exists($query, 'getValueBinder')
+                ? static::interpolate($sql, $query->getValueBinder()->bindings())
+                : static::interpolate($sql, $query->valueBinder()->bindings());
+        }
+
+        $trace = Debugger::trace(['start' => 1, 'depth' => $stackDepth + 2, 'format' => 'array']);
+        $file = $trace[$stackDepth]['file'];
+        $line = $trace[$stackDepth]['line'];
```

Comment:
You should ensure that the depth key is defined before using it.

Reply:
@markstory  You mean something like `$stackDepth = is_int($stackDepth) ? $stackDepth : 0`?

Reply:
No more like an `isse($trace[$stackDepth])`. This method is public and someone could send you `-10` as the depth.

Reply:
I think he means more the value behind the key: `$trace[$stackDepth]` could be undefined, right?

Reply:
oh right, that makes sense. I'll fix it.